File: tree-traversal/tree_traversal.go
package main

import (
	"fmt"
)

type Node struct {
	value int
	left  *Node
	right *Node
}

type BinarySearchTree struct {
	root *Node
}

func (bst *BinarySearchTree) Insert(value int) bool {
	newNode := &Node{value: value}
	if bst.root == nil {
		bst.root = newNode
		return true
	}
	temp := bst.root
	for {
		if newNode.value == temp.value {
			return false
		}
		if newNode.value < temp.value {
			if temp.left == nil {
				temp.left = newNode
				return true
			}
			temp = temp.left
		} else {
			if temp.right == nil {
				temp.right = newNode
				return true
			}
			temp = temp.right
		}
	}
}

func (bst *BinarySearchTree) BFS() []int {
	results := []int{}
	if bst.root == nil {
		return results
	}
	queue := []*Node{bst.root}

	for len(queue) > 0 {
		currentNode := queue[0]
		queue = queue[1:]
		results = append(results, currentNode.value)
		if currentNode.left != nil {
			queue = append(queue, currentNode.left)
		}
		if currentNode.right != nil {
			queue = append(queue, currentNode.right)
		}
	}
	return results
}

func (bst *BinarySearchTree) DFSPreOrder() []int {
	var results []int
	var traverse func(node *Node)
	traverse = func(node *Node) {
		results = append(results, node.value)
		if node.left != nil {
			traverse(node.left)
		}
		if node.right != nil {
			traverse(node.right)
		}
	}
	traverse(bst.root)
	return results
}

func (bst *BinarySearchTree) DFSPostOrder() []int {
	var results []int
	var traverse func(node *Node)
	traverse = func(node *Node) {
		if node.left != nil {
			traverse(node.left)
		}
		if node.right != nil {
			traverse(node.right)
		}
		results = append(results, node.value)
	}
	traverse(bst.root)
	return results
}

func (bst *BinarySearchTree) DFSInOrder() []int {
	var results []int
	var traverse func(node *Node)
	traverse = func(node *Node) {
		if node.left != nil {
			traverse(node.left)
		}
		results = append(results, node.value)
		if node.right != nil {
			traverse(node.right)
		}
	}
	traverse(bst.root)
	return results
}

func main() {
	tree := BinarySearchTree{}
	values := []int{28, 45, 98, 11, 21, 86, 21}
	for _, value := range values {
		tree.Insert(value)
	}

	fmt.Println("BFS:", tree.BFS())
	fmt.Println("DFS Pre-order:", tree.DFSPreOrder())
	fmt.Println("DFS Post-order:", tree.DFSPostOrder())
	fmt.Println("DFS In-order:", tree.DFSInOrder())
}

File: tree-traversal/tree_traversal.py
class Node:
    def __init__(self,value) :
        self.value =  value
        self.left = None
        self.right = None


class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self,value):
        new_node = Node(value)
        if self.root is None:
            self.root = new_node
            return True
        else:
            temp = self.root
            while (True):
                if new_node.value == temp.value:
                    return False
                if new_node.value < temp.value:
                    if temp.left is None:
                        temp.left = new_node
                        return True
                    temp = temp.left
                else:
                    if temp.right is None:
                        temp.right = new_node
                        return True
                    temp = temp.right

    def BFS(self):
        current_node = self.root
        queue = []
        results = []
        queue.append(current_node )

        while len (queue) > 0:
            current_node = queue.pop(0)
            results.append(current_node.value)
            if current_node.left is not None:
                queue.append(current_node.left)
            if current_node.right is not None:
                queue.append(current_node.right)
        return results
    
    def dfs_pre_order (self):
        results = []
        def traverse (current_node):
            results.append(current_node.value)
            if current_node.left is not None:
                traverse(current_node.left)
            if current_node.right is not None:
                traverse(current_node.right)
        traverse(self.root)
        return results
    
    def dfs_post_order(self):
        results = []    
        def traverse (current_node):
            
            if current_node.left is not None:
                traverse(current_node.left)
            if current_node.right is not None:
                traverse(current_node.right)
            results.append(current_node.value)
        traverse(self.root)
        return results
    
    def dfs_in_order(self):
        results = []    
        def traverse (current_node):
            
            if current_node.left is not None:
                traverse(current_node.left)
            results.append(current_node.value)
            if current_node.right is not None:
                traverse(current_node.right)
           
        traverse(self.root)
        return results

tree = BinarySearchTree()
tree.insert(28)
tree.insert(45)
tree.insert(98)
tree.insert(11)
tree.insert(21)
tree.insert(86)
tree.insert(21)

print("BFS result:", tree.BFS())



print("dfs pre order result:",tree.dfs_pre_order())

print("dfs post order result:",tree.dfs_post_order())

print("dfs in order result:",tree.dfs_in_order())